share/revng/test/tests/abi/runner.c

```diff
@@ -0,0 +1,226 @@
+ * 2 (third time) corresponds to the signal after the return.
+ * After which the state loops back to 0.
+ */
+int current_state = 0;
+
+/* Since it's not safe to call functions like `prinf` inside the handler (what
+ * if an interrupt happens while inside?), these memory locations are used as
+ * an intermediate place to store the state which can be printed later down
+ * the line
+ */
+uint8_t *stored_state[3];
+static void malloc_the_stored_state(void) {
+  uint64_t allocated_word_count = REGISTER_COUNT + input.stack_byte_count;
+  for (uint32_t i = 0; i < 3; ++i)
+    stored_state[i] = (uint8_t *) malloc(sizeof(size_t) * allocated_word_count);
+}
+static void free_the_stored_state(void) {
+  for (uint32_t i = 0; i < 3; ++i)
+    free(stored_state[i]);
+}
+
+void handler(int signal, siginfo_t *info, void *context) {
+  // Abort when an unexpected signal is raised just in case.
+  if (signal != SIGILL)
+    exit(4);
+
+  // Save registers.
+  uint8_t *current_storage = stored_state[current_state];
+  for (uint32_t i = 0; i < REGISTER_COUNT; ++i)
+    ((size_t *) current_storage)[i] = REGISTER(context, i);
+
+  // Save the stack.
+  memcpy(current_storage + REGISTER_COUNT * sizeof(size_t),
+         (void *) (size_t) SP(context),
+         input.stack_byte_count);
+
+  // Restore the instruction that caused current interrupt to its original value
+  if (current_state == 0)
+    ACCESS_A_WORD(saved.function->callsites[0].before) = saved.before_the_call;
+  else if (current_state == 1)
+    ACCESS_A_WORD(saved.function->address) = saved.after_the_call;
+  else if (current_state == 2)
+    ACCESS_A_WORD(saved.function->callsites[0].after) = saved.after_the_return;
+
+  if (++current_state > 2)
+    current_state = 0;
+}
+
+void print_a_register(const char *name, uint8_t *value);
+void print_stack(uint8_t *data);
+static void print_single_state(uint8_t *data) {
+  puts("      Registers:");
+  for (uint32_t i = 0; i < REGISTER_COUNT; ++i)
+    print_a_register(register_names[i], data + sizeof(size_t) * i);
+
+  printf("      Stack: ");
+  print_stack(data + sizeof(size_t) * REGISTER_COUNT);
+}
+
+static void print_saved_state(void) {
+  puts("    StateBeforeTheCall:");
+  print_single_state(stored_state[0]);
+  puts("    StateAfterTheCall:");
+  print_single_state(stored_state[1]);
+  puts("    StateAfterTheReturn:");
+  print_single_state(stored_state[2]);
+}
+
+void print_header(void);
+void print(const char *name, const uint8_t *input, const uint8_t *output);
+
+int main(int argc, char **argv) {
+  /* Setup signal handling */
+  struct sigaction action = { .sa_sigaction = &handler,
+                              .sa_flags = SA_SIGINFO };
+  sigemptyset(&action.sa_mask);
+  sigaction(SIGILL, &action, NULL);
+
+  if (argc != 2)
+    return 1;
+
+  /* load the "inner" binary */
+  int binary_file = open(argv[1], O_RDONLY);
+  struct stat binary_file_stats;
+  fstat(binary_file, &binary_file_stats);
```

Comment:
Assert return codes are those you expect of all POSIX calls.

Reply:
Done.